### ISSUE TITLE
Usuing Dataset and DataLoader

### ISSUE THREAD

**Issue by neojg** [dotnet/TorchSharp#828]
I'm trying to make a simple XOR example wit Dataset and Dataloader.

Please advise what is the reason of the error in the following code:

```C#
public class XORDataset : utils.data.Dataset
    {
        public XORDataset() {}

        private long size = 10;

        public override long Count => size;

        public override Dictionary<string, Tensor> GetTensor(long index)
        {
            switch (index % 4)
            {
                case 0:
                    return new Dictionary<string, Tensor>() { { "0", new int[2] { 0, 0 } } };
                case 1:
                    return new Dictionary<string, Tensor>() { { "1", new int[2] { 0, 1 } } };
                case 2:
                    return new Dictionary<string, Tensor>() { { "1", new int[2] { 1, 0 } } };
                case 3:
                    return new Dictionary<string, Tensor>() { { "0", new int[2] { 1, 1 } } };
            }
            return new Dictionary<string, Tensor>() { { "0", new int[2] { 1, 1 } } };
        }
    }

// Now I try to use it for training
// ...

                    foreach(var batch in dl)
// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
// The given key '1' was not present in the dictionary.
//    at System.ThrowHelper.ThrowKeyNotFoundException[T](T key)
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at TorchSharp.torch.utils.data.DataLoader.<>c__DisplayClass2_0.<Collate>b__0(Dictionary`2 k)
   at System.Linq.Enumerable.SelectListIterator`2.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at TorchSharp.torch.utils.data.DataLoader.Collate(IEnumerable`1 dic, Device device)
   at TorchSharp.torch.utils.data.DataLoader`2.DataLoaderEnumerator.MoveNext()
   at Arhat.InvoiceDigitizer.AI.Class1.TestDL() in C:\Work\NumeryFaktur\AI\Class1.cs:line 114
   at ConsoleApp1.Program.Main(String[] args) in C:\Work\NumeryFaktur\ConsoleApp1\Program.cs:line 22
//
                    {
                        foreach(var data in batch)

                        {
                            using (var eval = seq.forward(data.Value))
                            using (var output = functional.mse_loss(eval, data.Key =="T" ? 1 : 0, Reduction.Sum))
                            {
                                optimizer.zero_grad();
                                output.backward();
                                optimizer.step();

                                var loss = output.data<float>()[0];
                                Console.WriteLine($"Epoch: {i} Loss: {loss}");
                            }
                        }
                    }
```

The exception "The given key '1' was not present in the dictionary." is risen at the first iteration of getting the batch from dl (DataLoader).

What am I missing??

**Comment by neojg** [comment]
I hope I got the idea.

The error was in the point that the string in Dictionary<string, Tensor> is the label of the features and should be the same in each sample set like each dictionary be the structure { data: Tensor, value: Tensor, ... }

corrected dtataset:

```C#
public override Dictionary<string, Tensor> GetTensor(long index)
        {
            switch (index % 4)
            {
                case 0:
                    return new Dictionary<string, Tensor>()
                    {
                        { "data", new float[2] { 0, 0 } },
                        { "result", new float[1] { (float) 0.0 } }
                    };
                case 1:
                    return new Dictionary<string, Tensor>()
                    {
                        { "data", new float[2] { 0, 1 } },
                        { "result", new float[1] { (float) 1.0 } }
                    };
                case 2:
                    return new Dictionary<string, Tensor>()
                    {
                        { "data", new float[2] { 1, 0 } },
                        { "result", new float[1] { (float) 1.0 } }
                    };
                case 3:
                    return new Dictionary<string, Tensor>()
                    {
                        { "data", new float[2] { 1, 1 } },
                        { "result", new float[1] { (float) 0.0 } }
                    };
            }
            return new Dictionary<string, Tensor>()
                    {
                        { "data", new float[2] { 0, 0 } },
                        { "result", new float[1] { (float) 0.0 } }
                    };
        }
```
